Initial draft of the SSSOM/RDF spec. (#469)
mapping-commons/sssom: src/docs/spec-formats-rdf.md

#### For slots typed as `linkml:date`
(e.g. `mapping_date`, `publication_date`)

The value is represented as a `xsd:date` literal.

#### For slots typed as `linkml:double`
(e.g. `mapping_set_confidence`, `confidence`, `similarity_score`)

The value is represented as a `xsd:double` literal.

#### For slots typed as an enumeration
(e.g. `sssom_version`, `mapping_cardinality`, `subject_type`…)

If the permissible values for the enumeration are defined in the LinkML
model as having an associated `meaning` property, then the value is
represented as a named RDF resource with the indicated property.
Otherwise, the value is represented as a string literal.

#### For slots typed as a SSSOM object
(e.g. `mappings`, `extension_definitions`)

The value is represented as a RDF resource. Whether the resource is
named (IRI) or not (blank node) will depend on the type of the object,
see the [section on representing SSSOM objects](#sssom-objects) below
for details.

### Representation of multi-valued slots
(e.g. `creator_id`, `see_also`, `object_match_field`…)

As an exception to the general principle that slots are represented by a
single RDF triple, multi-valued slots are represented by as many

> triples as there are values, each value being the object of one triple.

Author: > for each value {v1,v..,vn} represented by a set of individual triples {a,b,v1; a,b,v2,...a,b,vn}.

Does not sound any clearer to me. On the contrary, it sounds like each value is represented by a set of triples, which is certainly not the case.

> Also, is this true for mappings slot as well? Probably not right?

Of course it is. Mappings are represented as follows:

```ttl
?mappingset sssom:mappings [ a owl:Axiom ;
                               owl:annotatedSource ...
                           ] ,
                           [ a owl:Axiom ;
                               owl:annotatedSource ...
                           ] .
```

which fits the description for multi-valued slots: one triple per value.

This is what SSSOM-Py has always done, so I had assumed you were fine with that.

Let me guess: you are no longer happy with that and want to radically change the format?

> triples as there are values, each value being the object of one triple.

Reviewer: No, it seems I misunderstood  

```
multi-valued slots are represented by as many
triples as there are values, each value being the object of one triple.
```

I thought this literally meant:

```
for each value {v1,v..,vn} represented by a set of individual triples {a,b,v1; a,b,v2,...a,b,vn}.
```

Which is why I was confused.

Maybe just add an example here?

> Non-normative notes:
>
> 1. This means, in particular, that RDF complex structures intended to
>    represent collections of values, such as `rdfs:Container` or
>    `rdfs:List`, MUST NOT be used to represent multi-valued SSSOM
>    slots.
> 2. This also implies that values in multi-valued slots are _not_
>    ordered.

The other rules above apply to determine how each single value is to be
represented.

<a id="extension-slots"></a>
### Representation of extension slots
An [extension slot](spec-model.md#non-standard-slots) is represented in
a similar way to a standard slot, with the following specific rules.

The predicate is the property associated to the extension slot, as
indicated by the `property` slot in the set’s
[definition](ExtensionDefinition.md) of the extension.

The value of the extension is represented:

* as a named RDF resource, if the `type_hint` of the extension
  definition is `linkml:uriOrCurie`;

> * otherwise, as a literal of the type indicated by the `type_hint`.

Author: First, I would strongly advise against using `sssom:NonRelativeURI` as the `type_hint` for an extension slot.

That type is a hack that is only needed because (1) some people have strong opinion on what a `xsd:anyURI` value should look like (in particular, it should not store a relative URI reference) and (2) SSSOM tends to use URI-typed slots where an EntityReference would have been preferable given the intended usage of the concerned slots.

I would very much like this hack to remain purely internal to SSSOM. If people have URI-typed values, then

* either those values are semantic identifiers, and they should use `linkml:uriOrCurie`;
* or those values are merely, well, URIs, and there should be no reason for them not to use `xsd:anyURI`. Everybody else outside of SSSOM has been happy with that type for the past 20 years.

Using a `sssom:NonRelativeURI` type in the RDF representation of a mapping set is likely to do more harm than good, if only because, outside of SSSOM, nobody will have the slightest idea of what that type is – whereas any RDF person is likely to be familiar with `xsd:anyURI`.

Second: Simplicity and avoidance of special cases. It’s much easier, both for the spec writer and for the spec implementer, if all extension values can be treated the same way. There is _one_ special case (`uriOrCurie`) for semantic identifiers, but beyond that, the value of the `type_hint` does not matter.

## Representation of SSSOM objects

### Representation of a `Mapping` object
The RDF type of a `Mapping` object is `owl:Axiom`.

If the `Mapping` object has a `record_id` slot, then the value of that
slot is used as the named RDF resource that represents the object (and
consequently, that slot MUST NOT be represented using the [general
rules](#sssom-slots) for the representation of slots as defined above).
Otherwise, the `Mapping` object is represented as a blank node.

### Representation of a `MappingSet` object
The RDF type of a `MappingSet` object is `sssom:MappingSet`.

A `MappingSet` object is represented by a named RDF resource
corresponding to the value of the `mapping_set_id` slot (which
consequently MUST NOT be represented using the [general
rules](#sssom-slots) for the representation slots as defined above).

The `curie_map` slot MUST NOT be represented using the [general
rules](#sssom-slots). Instead, if it is needed it MUST be represented
using whatever mechanism is provided by the concrete RDF serialisation
format (e.g. `@prefix` declarations in [RDF
Turtle](https://www.w3.org/TR/turtle/) or [RDF
TriG](https://www.w3.org/TR/trig/), or `xmlns` namespace declarations in

> [RDF/XML](https://www.w3.org/TR/rdf-syntax-grammar/)).

Reviewer: > You are joking. You are joking, right?

I did not expect such a strong disagreement :P 

Alright, instead of this suggestion, we should then make SSSOM-RDF explicitly write only format.. I am totally fine with this as well, and we don't need to concern ourselves then _at all_ with the curie_map.. This way we can circumvent the limitations outlined above?

> [RDF/XML](https://www.w3.org/TR/rdf-syntax-grammar/)).

Author: > I did not expect such a strong disagreement :P

Try sparing some thoughts for the people who have to implement your last-minute ideas that you sprout out of nowhere, maybe you’ll understand.

Two days ago you were reluctant to just changing the predicates to use to represent the core triple (e.g. changing from `owl:annotatedSource` to `sssom:subject_id`, which would have been a much smaller change and easier change than what you’re suggesting here). I thought we were in agreement that the SSSOM/RDF format should be as close as possible to what we already have, both to avoid needlessly breaking things and to keep the work needed to make SSSOM-Py support the updated format minimal (given that we have very little developer time available to work on SSSOM-Py). The spec I proposed is tailored for that; it deviates only minimally from the existing, _de facto_ SSSOM/RDF format that SSSOM-Py has been producing since forever; it is already completely supported by SSSOM-Java and I am hopeful that SSSOM-Py could be made to support it relatively easily (precisely due to the minimal deviations). You were fine with the [initial proposal](https://github.com/mapping-commons/sssom/discussions/454#discussioncomment-13655841), on which you said that you had merely “minor stylistic comments”.

And now, you’re here suggesting that we should in fact do something completely different, that has never even been casually mentioned in any discussion related to the SSSOM/RDF format. Because all of a sudden you are concerned about roundtripping back from RDF/XML, even though the RDF/XML produced by SSSOM-Py has never been roundtrippable and AFAIK nobody has ever complained about that.

So yeah, I disagree with your proposal, because it contradicts everything that you _seemed_ to want just two days ago.

So what do you want _now_?

A. A SSSOM/RDF format that is minimally different from what we already have, that can be supported rapidly (that is _already_ supported by one implementation), but that (oh, the horror!) does not guarantee that a set written in RDF/XML can be roundtripped back to another SSSOM format?

B. A SSSOM/RDF format that is a clean break from the existing stuff, that will initially not be supported by any implementation (and in fact I doubt it will _ever_ be implemented by SSSOM-Py, given the lack of activity on that implementation)?

If you want B, fine. But then I’ll leave _you_ to design the format. I won’t get involved in any of it, I’ll just wait until you have designed the perfect format of your dream, and then I promise I’ll do my best to implement it.

> [RDF/XML](https://www.w3.org/TR/rdf-syntax-grammar/)).

Author: > we should then make SSSOM-RDF explicitly write only format.. 

Now you are throwing the baby with the bathwater. Just because the RDF/XML concrete serialisation may not guarantee that the prefix map is preserved does not mean that we should give up on SSSOM/RDF being a read/write format.

Again: As currently written, the spec does allow a SSSOM/RDF set to be fully converted back to any other SSSOM format, provided that:

* you do not serialise into RDF/XML;
* you serialise identifiers as CURIEs and make sure to include the appropriate prefix declarations.

As I said in another comment below, I wrote the spec to be flexible (“à la carte”): if you want the ability to roundtrip between RDF and another format, you can have it; if you are not interested in that ability, you can ignore it.

> [RDF/XML](https://www.w3.org/TR/rdf-syntax-grammar/)).

Author: > * you do not serialise into RDF/XML;
> * you serialise identifiers as CURIEs and make sure to include the appropriate prefix declarations.

In fact even if you do serialise into RDF/XML and write identifiers as IRIs, you will _still_ be able convert back to SSSOM/TSV, _unless_ you happen to run your RDF/XML file into a RDF processor that decides to strip away any unused namespace declarations. Not sure if that is a common behaviour among RDF tools, but Jena and RDFLib do _not_ seem to do it – they are happy to let unused namespace declarations pass through unchanged.

> Non-normative notes
>
> 1. The CURIE map may not be needed at all if all named resources and

> >    predicates are always serialised as full-length IRIs.

Author: This is why there is a subsection ”serialisation of identifiers” in the “Special considerations” section.

The SSSOM/RDF format can be _both_ a read/write format that is equivalent to SSSOM/TSV or SSSOM/JSON (meaning that can roundtrip between all those formats without loss of information) _and_ an export format. It all depends on what you want to do with the output file – something the spec cannot know in advance.

In a sense, SSSOM/RDF is a “à la carte” format. You want to preserve the ability to roundtrip back to SSSOM/TSV? You can, just make sure to include the CURIE map and any extension definitions. You are not interested in being able to come back (say, because all you want is to load the set into a graph database and forget about it)? Then you don’t have to worry about the CURIE map or extension definitions at all.

> 2. If at least some named resources or predicates are serialised as
>    CURIEs, the RDF requirement that all used prefix names must be
>    declared (using the appropriate mechanism for the chosen concrete
>    syntax) takes precedence over the possibility of omitting the
>    declarations of prefix names that are considered

> >    [built-in](spec-intro.md#iri-prefixes) in the context of SSSOM.

Reviewer: I dont know what my July-self exactly was thinking here.

### Representation of a `ExtensionDefinition` object
The RDF type of a `ExtensionDefinition` object is
`sssom:ExtensionDefinition`.

A `ExtensionDefinition` object has no identifier of any kind and is
always represented by a blank node.

## Special considerations for serialising to RDF
When serialising a mapping set to SSSOM/RDF, implementations should
consider how the resulting RDF file is intended to be used. In
particular, they should ponder whether it is expected that the RDF
serialisation can at any time be converted back to any other SSSOM
format (e.g. SSSOM/TSV), or if it is only intended to be used by
“generic”, non-SSSOM-aware RDF applications.

Depending on that intended usage (if it is known), implementations may
adopt slightly different behaviours as described in the following
subsections.

### Serialisations of identifiers
If the serialisation is intended to be convertible back to another SSSOM

> format (especially the SSSOM/TSV format), implementations MUST

Author: Start by stating exactly what you expect from the SSSOM/RDF spec. I thought I knew, but in fact I cannot know it since it seemingly changes every other day.

> application will not know to propagate the condensed slots at the set
> level down to the level of the individual mappings, which will result
> in the application having an incomplete view of the mappings.

## Compatibility with pre-standard RDF representations
The present specification of the SSSOM/RDF format differs slightly from
what several implementations of SSSOM have been producing before the
format was formally specified.

In the name of backward compatibility, implementations MAY support the
alternative rules described in the following subsections when
deserialising from RDF.

Implementations MUST NOT follow these rules when serialising to RDF.

### Representation of slots typed as `sssom:NonRelativeURI`
Implementations MAY accept a value represented as a `xsd:anyURI`
literal.

### Representation of slots typed as an enumeration
Implementations MAY accept a value represented as a string literal, even
if the value is defined in the LinkML model as having an associated
`meaning` property.

For example, implementations MAY accept

```ttl
?mapping sssom:predicate_modifier "Not"^^xsd:string .
```

as an alternative to

```ttl

> ?mapping sssom:predicate_modifier sssom:NegatedPredicate .

Reviewer 2: why not just decide on one and say "this is the standard"?

Author: That’s what did. This:

```ttl
?mapping sssom:predicate_modifier sssom:NegatedPredicate .
```

**is** the standard.

But the decision to standardize that form has only been made a few weeks ago. Before that, both SSSOM-Java and SSSOM-Py have been producing the string literal form (SSSOM-Java for the past 8 months – since version 1.1, which introduced RDF support – and SSSOM-Py for as long as it has existed). In fact SSSOM-Py _still_ produces the string literal form to this day.

So for backwards compatibility (which is the entire point of this section, “compatibility with pre-standard representations”), implementations MAY support the old string literal form, even though it is _not_ the standard form.

Reviewer: Makes sense.

  dcterms:license <https://creativecommons.org/licenses/by/4.0/>;
  sssom:extension_definitions [
      sssom:property EXT:isFooable;
      sssom:slot_name "ext_fooable";
      sssom:type_hint xsd:boolean
    ];
  sssom:mappings [ a owl:Axiom;
      pav:authoredBy ORCID:0000-0002-7356-1779;
      dcterms:created "2025-07-14"^^xsd:date;
      owl:annotatedProperty skos:exactMatch;
      owl:annotatedSource KF_FOOD:F001;
      owl:annotatedTarget FOODON:00002473;
      EXT:isFooable true;
      sssom:confidence 9.5E-1;
      sssom:mapping_justification semapv:ManualMappingCuration;
      sssom:object_label "apple (whole)";
      sssom:subject_label "apple"
    ], [ a owl:Axiom;
      pav:authoredBy ORCID:0000-0002-7356-1779;
      dcterms:created "2025-07-14"^^xsd:date;
      owl:annotatedProperty skos:exactMatch;
      owl:annotatedSource KF_FOOD:F002;
      owl:annotatedTarget FOODON:00003348;
      EXT:isFooable false;
      sssom:confidence 1.0E0;
      sssom:mapping_justification semapv:ManualMappingCuration;
      sssom:object_label "Gala apple (whole)";
      sssom:subject_label "gala"
    ] .

> ```

Author: They appear, but only as reified OWL axioms.

This has been the RDF output produced by SSSOM-Py since the beginning.

> ```

Reviewer: Its a valid question though. Without the direct triple, triples stores might not be able to return all terms mapped to `?x` with a simple triple pattern - this is only possible because in many of the pipelines I had build that are doing gathering dust, I loaded the sssom file into robot and saved it, which I believe automatically injects that triple. 

In the OWL serialisation it seems I have [injected them](https://github.com/mapping-commons/sssom-py/blob/ab69eeb9f0611d8e5d0c436f4ecf556ded8bea97/src/sssom/writers.py#L215) in sssom py:
https://github.com/mapping-commons/sssom-py/blob/master/tests/validate_data/cob-to-external.tsv.owl

> ```

Reviewer: I am now on board with:

> I wouldn’t mind adding that as a OPTIONAL behaviour for RDF writer, on the condition that it is really optional – that is, if those ?subject_id ?predicate_it ?object_id triples are absent the set must still be accepted by a SSSOM/RDF reader, as long as the ?mapping owl:annotatedSource ?subject_id ; owl:annotatedPredicate ?predicate_id ; owl:annotatedTarget ?object_id triples are present.

With regards to the special cases:

> I’d say, we should do for them the same thing as we do for negated mappings (the two possibilities outlined in my previous message).

I agree. Just because both are very different use cases I would favour it if implementations would be injecting triples to sssom:NoTermFound versus negated mappings based on _separate_ conditionals; I would probably never add either one, but there may be some use cases to do so.

> ```

Author: > I would favour it if implementations would be injecting triples to sssom:NoTermFound versus negated mappings based on separate conditionals

Please define what you mean by “favour”. For now, we are talking about a completely optional behaviour, as in:

> A SSSOM/RDF writer MAY additionally inject for every mapping record a triple of the form ?subject_id ?predicate_id ?object_id.

(Note the _MAY_.) Since writing the SPO triple is an optional behaviour, we cannot expect any “favour” from the implementations, which are completely free to (1) implement the behaviour or not and (2) to implement it in whatever way they want.

If you _want_ implementations to offer the option of injecting SPO triples, then you have to _require_ it in the spec, as in:

> SSSOM/RDF writers MUST allow to optionally inject for every mapping record a triple of the form ?subject_id ?predicate_it ?object_id.

Note the difference compared to the previous phrasing. Here it’s no longer an _optional feature_, it’s a _mandatory_ feature (whose _activation_ is optional, but the feature itself has to be present, it’s no longer left for implementers to decide whether they want to support it or not).

> ```

Author: If we add to that the issue about `sssom:NoTermFound` and negated mappings, in the end you seem to want to require _three_ options:

(1) One option to enable/disable the production of SPO triple entirely:

> SSSOM/RDF writers MUST allow to optionally inject for every mapping record a triple of the form ?subject_id ?predicate_id ?object_id.

(2) One option to enable the production of SPO triples to `sssom:NoTermFound`:

> When the injection of ?subject_id ?predicate_id ?object_id triples is enabled, SSSOM/RDF writers MUST NOT inject such triples for records where the subject ID or the object ID (or both) is `sssom:NoTermFound`, unless this has been explicitly requested.

(3) One option to enable the production of SPO triples for negated mappings:

> When the injection of ?subject_id ?predicate_id ?object_id triples is enabled, SSSOM/RDF writers MUST NOT inject such triples for records where `predicate_modifier` is set to `Not`, unless this has been explicitly requested.

Frankly this seems way overkill to me. You even say yourself that you would probably never have a need for such options, so why _requiring_ from implementations that they offer them?

> ```

Author: > Ok, I agree with the plan. […] I would keep your phrasing of (2) and (3)

Err, with _what_ do you agree? In the end, you still want the _three_ options.

> ```

Reviewer: I thought you meant to make implementing these options an optional possibility? Maybe state again what exactly you are in favour of documenting here. I thought you were just against "requiring" these options? I am just saying that:

1. If you add that option, dont just inject noTermFound triples by default

> ```

Reviewer: I will resolve this comment now, looks great

      `https://w3id.org/sssom/` namespace and the name of the slot;
* the object is the value of the slot.

### Representation of slot values
The following rules determine how the value of a slot is represented as
the object of a RDF triple.

#### For slots typed as `sssom:EntityReference`
(e.g. `subject_id`, `mapping_justification`, `subject_source`…)

The value MUST be represented as a named RDF resource (IRI).

#### For slots typed as `sssom:NonRelativeURI`
(e.g. `license`, `mapping_provider`, `issue_tracker`…)

The value MUST be represented as a named RDF resource (IRI).

#### For slots typed as `linkml:date`
(e.g. `mapping_date`, `publication_date`)

The value MUST be represented as a `xsd:date` literal.

#### For slots typed as `linkml:double`
(e.g. `mapping_set_confidence`, `confidence`, `similarity_score`)

The value MUST be represented as a `xsd:double` literal.

#### For slots typed as an enumeration
(e.g. `sssom_version`, `mapping_cardinality`, `subject_type`…)

If the permissible values for the enumeration are defined in the LinkML
model as having an associated `meaning` property, then the value MUST be
represented as a named RDF resource with the indicated property.

> Otherwise, the value MUST be represented as a string literal.

Reviewer: ```suggestion
Otherwise, the value MUST be represented as an `xsd:string` literal.
```

To maintain consistency with the descriptions above

Author: Doesn’t seem strictly necessary to me as it should be obvious (when we are talking about RDF) that a “string literal” is a `xsd:string` literal, but I can’t object to the consistency argument. :)

The “**an**” before the `xsd:string` looks very weird to me though, and for consistency with other instances of “a `xsd:...`” I’m going to reject it.

(e.g. `mappings`, `extension_definitions`)

The value MUST be represented as a RDF resource. Whether the resource is
named (IRI) or not (blank node) will depend on the type of the object,
see the [section on representing SSSOM objects](#sssom-objects) below
for details.

### Representation of multi-valued slots
(e.g. `creator_id`, `see_also`, `object_match_field`…)

As an exception to the general principle that slots are represented by a
single RDF triple, multi-valued slots MUST be represented by as many
triples as there are values, each value being the object of one triple.

> Non-normative notes:
>
> 1. This means, in particular, that RDF complex structures intended to
>    represent collections of values, such as `rdfs:Container` or
>    `rdfs:List`, MUST NOT be used to represent multi-valued SSSOM
>    slots.
> 2. This also implies that values in multi-valued slots are _not_
>    ordered.

The other rules above apply to determine how each single value is to be
represented.

> Example:
>
> A `creator_id` slot with the values `https://example.org/people/0001`
> and `https://example.org/people/0002` is represented by the following
> two triples:
>
> ```ttl

> > ?object pav:authoredBy <https://example.org/people/0001> .

Reviewer: ```suggestion
> ?object dcterms:creator <https://example.org/people/0001> .
```

Reviewer: https://mapping-commons.github.io/sssom/creator_id/

> > ?object pav:authoredBy <https://example.org/people/0002> .

Reviewer: ```suggestion
> ?object dcterms:creator <https://example.org/people/0002> .
```

    slot, if such a field is present;
  - or a property constructed by concatenating the `https://w3id.org/sssom/`
    namespace and the name of the slot;
- the object is the value of the slot.

### Representation of slot values

The following rules determine how the value of a slot is represented as the
object of a RDF triple.

#### For slots typed as `sssom:EntityReference`

(e.g. `subject_id`, `mapping_justification`, `subject_source`…)

The value MUST be represented as a named RDF resource (IRI).

#### For slots typed as `sssom:NonRelativeURI`

(e.g. `license`, `mapping_provider`, `issue_tracker`…)

The value MUST be represented as a named RDF resource (IRI).

#### For slots typed as `linkml:date`

(e.g. `mapping_date`, `publication_date`)

The value MUST be represented as a `xsd:date` literal.

#### For slots typed as `linkml:double`

(e.g. `mapping_set_confidence`, `confidence`, `similarity_score`)

The value MUST be represented as a `xsd:double` literal.

#### For slots typed as an enumeration

(e.g. `sssom_version`, `mapping_cardinality`, `subject_type`…)

If the permissible values for the enumeration are defined in the LinkML model as

> having an associated `meaning` property, then the value MUST be represented as a

Reviewer 2: can you please give an explicit example?

named RDF resource with the indicated property. Otherwise, the value MUST be
represented as a `xsd:string` literal.

#### For slots typed as a SSSOM object

> (e.g. `mappings`, `extension_definitions`)
> 
> The value MUST be represented as a RDF resource. Whether the resource is named
> (IRI) or not (blank node) will depend on the type of the object, see the
> [section on representing SSSOM objects](#sssom-objects) below for details.

Reviewer 2: this is a bit confusing why it's documented here but not in full, can you make an explicit enumeration here of the three things this applies to an an link to each?

```suggestion
The value MUST be represented as a RDF resource. Whether the resource is named
(IRI) or not (blank node) will depend on the type of the object, see the following sections:

1. [mappings](#representation-of-a-mapping-object)
2. [mapping set)(#representation-of-a-mappingset-object)
3. [extension definition](#representation-of-a-extensiondefinition-object)
```

When serialising a mapping set to SSSOM/RDF, implementations should consider how
the resulting RDF file is intended to be used. In particular, they should ponder
whether it is expected that the RDF serialisation can at any time be converted
back to any other SSSOM format (e.g. SSSOM/TSV), or if it is only intended to be
used by “generic”, non-SSSOM-aware RDF applications.

Depending on that intended usage (if it is known), implementations may adopt
slightly different behaviours as described in the following subsections.

### Serialisations of identifiers

If the serialisation is intended to be convertible back to another SSSOM format
(especially the SSSOM/TSV format), implementations MUST declare all the prefixes
found in the CURIE map and SHOULD serialise all identifiers as CURIEs using said
declared prefixes.

> Non-normative explanation
>
> This is because, if all identifiers are serialised as full-length IRIs, then
> even if the RDF file includes prefix declarations, they may be stripped away
> by a RDF reader, since they are not needed. And without those prefix
> declarations, it would not be possible to serialise the set back as a

> > SSSOM/TSV file (remember that the SSSOM/TSV format _requires_ that identifiers
> > be serialised as CURIEs).

Reviewer 2: shade thrown at CURIEs :p. Let's extend this with a sentence with some context like 

```suggestion
> SSSOM/TSV file (remember that the SSSOM/TSV format _requires_ that identifiers
> be serialised as CURIEs to support the easy understanding on its content).
```

Author: Sorry, what? This is a factual statement, how is that a “shade thrown at CURIEs”?

And I believe this factual statement is necessary because people regularly forget the SSSOM/TSV-specific requirement that identifiers MUST be in CURIE form – even the very people who should know very well about that requirement, as in [this discussion](https://github.com/mapping-commons/sssom/issues/399#issuecomment-2493913177).

Besides, your added sentence is incorrect. The fact that SSSOM/TSV absolutely requires identifiers to be in CURIE form has nothing to do with “supporting the easy understanding on its contents”.

First because it is dubious that `FBbt:00004508` is easier to understand that `http://purl.obolibrary.org/obo/FBbt_00004508` – personally I’d say it’s _worse_, because know you need to look up what `FBbt` stands for, whereas the full-length IRI is self-sufficient.

Second, even if CURIEs were actually easier to understand, SSSOM/TSV could very well have simply _encouraged_ the use of CURIE forms, instead of _requiring_ it (as in, “identifiers in SSSOM/TSV files SHOULD be CURIE form”, instead of “identifiers in SSSOM/TSV files MUST be in CURIE form”).

The real reason SSSOM/TSV absolutely requires the use of CURIEs is, if I recall correctly, because SSSOM-Py developers didn’t want to have to deal with the possibility that a SSSOM/TSV could contain both full-length identifiers and CURIEs. Notably because this would have required, either the use of some heuristics to infer the form of an identifier, or the use of some syntactic sugar to distinguish between an IRI and a CURIE, like what is done in most RDF syntaxes – e.g. IRI in angled brackets vs “naked” CURIEs –, which nobody wanted.

But regardless all that, there is no need for the RDF specification to elaborate on _why_ the SSSOM/TSV requires the use of CURIEs. People simply need to be reminded that it is the case, because they frequently forget about it and this has some consequences if you want to be able to re-convert a SSSOM/RDF mapping set back to the SSSOM/TSV format.

  set to `Not`);
- when the record represents an absence of match (that is, `subject_id` or
  `object_id` – or both – is set to `sssom:NoTermFound`).

In any case, a SSSOM/RDF reader MUST NOT expect the presence of such triples,
and if they are present MUST NOT use them to construct mapping records.

> Non-normative explanations
>
> Such “direct triples” are merely a convenience for downstream RDF
> applications, allowing them to find a direct link (as a single triple) between
> the subject and the object of a mapping, without having to construct such a
> link by following the `owl:annotatedSource`, `owl:annotatedProperty`, and
> `owl:annotatedTarget` triples.
>
> It is recommended not to inject such direct triples for literal mapping
> records, even if they do have a `subject_id` and a `object_id`, because by
> definition the subject and/or the object of such records is not an
> identifiable semantic entity and has no business being represented in a RDF
> graph.
>
> It is recommended not to inject such direct triples for negated mapping
> records because they would seem to convey a meaning that is the exact opposite
> of what the records mean.
>

> > It is recommended not to inject such direct triples for no-match mapping

Reviewer 2: I meant making our own RDF idiom for it

# The SSSOM/RDF serialisation format

This section defines how to represent a SSSOM mapping set as a
[RDF model](https://www.w3.org/TR/rdf11-concepts/).

## RDF formats

The RDF model that represents a SSSOM mapping set is independent of the concrete
format that may be used to serialise the model.

It is RECOMMENDED that implementations support reading and writing a SSSOM set
from and to the [RDF Turtle](https://www.w3.org/TR/turtle/) format at least.
They MAY support any other RDF concrete format (e.g. RDF/XML, TriG, N-Triples,
etc.).

This specification does not mandate how a concrete RDF syntax is to be used. For
example, if the RDF syntax allows named resources and predicates to be

> serialised as either IRIs or CURIEs, if is left at the discretion of the

Reviewer 3: ```suggestion
serialised as either IRIs or CURIEs, it is left to the discretion of the
```

as defined above).

The `curie_map` slot MUST NOT be represented using the
[general rules](#sssom-slots). Instead, if it is needed it MUST be represented
using whatever mechanism is provided by the concrete RDF serialisation format
(e.g. `@prefix` declarations in [RDF Turtle](https://www.w3.org/TR/turtle/) or
[RDF TriG](https://www.w3.org/TR/trig/), or `xmlns` namespace declarations in
[RDF/XML](https://www.w3.org/TR/rdf-syntax-grammar/)).

> Non-normative notes
>
> 1. The CURIE map may not be needed at all if all named resources and
>    predicates are always serialised as full-length IRIs.
> 2. If at least some named resources or predicates are serialised as CURIEs,
>    the RDF requirement that all used prefix names must be declared (using the
>    appropriate mechanism for the chosen concrete syntax) takes precedence over
>    the possibility of omitting the declarations of prefix names that are
>    considered [built-in](spec-intro.md#iri-prefixes) in the context of SSSOM.

### Representation of a `ExtensionDefinition` object

The RDF type of a `ExtensionDefinition` object is `sssom:ExtensionDefinition`.

> A `ExtensionDefinition` object has no identifier of any kind and is always

Reviewer 3: ```suggestion
### Representation of an `ExtensionDefinition` object

The RDF type of an `ExtensionDefinition` object is `sssom:ExtensionDefinition`.

An `ExtensionDefinition` object has no identifier of any kind and is always
```

inject a single triple of the form:

```ttl
?subject_id ?predicate_id ?object_id .
```

If so, that behaviour MUST be optional.

When that behaviour is enabled, implementations SHOULD NOT inject such triples
in the following cases:

- when the record represents a literal mapping (that is, `subject_type` or
  `object_type` – or both – is set to `rdfs literal`);
- when the record represents a negated mapping (that is, `predicate_modifier` is
  set to `Not`);
- when the record represents an absence of match (that is, `subject_id` or
  `object_id` – or both – is set to `sssom:NoTermFound`).

In any case, a SSSOM/RDF reader MUST NOT expect the presence of such triples,
and if they are present MUST NOT use them to construct mapping records.

> Non-normative explanations
>
> Such “direct triples” are merely a convenience for downstream RDF
> applications, allowing them to find a direct link (as a single triple) between
> the subject and the object of a mapping, without having to construct such a
> link by following the `owl:annotatedSource`, `owl:annotatedProperty`, and
> `owl:annotatedTarget` triples.
>
> It is recommended not to inject such direct triples for literal mapping

> > records, even if they do have a `subject_id` and a `object_id`, because by

Reviewer 3: ```suggestion
> records, even if they do have a `subject_id` and an `object_id`, because by
```
